### PR TITLE
Avoid hang in preClose method during dup2 system call

### PR DESCRIPTION
**Pull request by shruacha1234**
During fcntl, dup2 system calls if read/write happens then PreClose method gets hang. This issue is not seen if we Signal/Kill the thread first and then call the preClose method.

### CODE REVIEW

**Comment by mstoodle** [comment]
probably not relevant and there's probably a reason for it, but it's interesting that NioSocketImpl.close() will call preClose(fd) even if reader==0 && writer==0 but all the other callers only call preclose(fd) conditionally if reader!=0 || writer!=0 (or if thread!=0)

File: src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java
Comment (pshipton) [inline src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java:1741]: This needs to be concerned with the SecurityManager, so it should use `GetPropertyAction.privilegedGetProperty`. Also it shouldn't fetch "os.name" every time the method is called, but do it once and cache the result.

For an example see https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/src/java.base/share/classes/sun/security/provider/SunEntries.java#L98

Comment (pshipton) [reply src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java:1741]: You could add a new public class sun.net.util.AIX and define a `public static final boolean isAix` variable there for every place to use.

**Comment by keithc-ca** [comment]
It's not clear to me that this _guarantees_ to avoid the problem. If an interrupted thread immediately re-issues its read or write call (a likely scenario I believe) it may already be blocked there (again) before the dup2 call occurs.

File: src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java
Comment (keithc-ca) [inline src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java:1755]: Rather than repeating as much code, suggest making the call to `preClose()` conditional before or after the `signal()` calls:
```java
                    if (!AIX.isAIX)
                        nd.preClose(fd);
                    if (reader != 0)
                        NativeThread.signal(reader);
                    if (writer != 0)
                        NativeThread.signal(writer);
                    if (AIX.isAIX)
                        nd.preClose(fd);
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by patilca** [comment]
@shruacha1234 Will you please be able to address Peter's review comments regarding SecurityManager? He will need those before launching PR testing. Thanks

**Comment by pshipton** [comment]
jenkins test sanity,extended.functional,sanity.system,extended.system aix jdk17

File: closed/src/java.base/share/classes/sun/net/util/AIX.java
Comment (keithc-ca) [inline closed/src/java.base/share/classes/sun/net/util/AIX.java:29]: Please delete blank line 24 and indent four spaces per level (like the rest of this repo), not three.
This is missing:
```java
import sun.security.action.GetPropertyAction;
```

File: src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java
Comment (keithc-ca) [inline src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java:1741]: `AIX` must be imported.

File: src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java
Comment (keithc-ca) [inline src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java:32]: Just one blank line, please.
This is missing:
```java
import sun.net.util.AIX;
```

File: src/java.base/share/classes/sun/nio/ch/ServerSocketChannelImpl.java
Comment (keithc-ca) [inline src/java.base/share/classes/sun/nio/ch/ServerSocketChannelImpl.java:592]: Missing import for `AIX`.

File: src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java
Comment (keithc-ca) [inline src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java:1023]: Missing import for `AIX`.

File: src/java.base/unix/classes/sun/nio/ch/SinkChannelImpl.java
Comment (keithc-ca) [inline src/java.base/unix/classes/sun/nio/ch/SinkChannelImpl.java:132]: Missing import for `AIX`.

File: src/java.base/unix/classes/sun/nio/ch/SourceChannelImpl.java
Comment (keithc-ca) [inline src/java.base/unix/classes/sun/nio/ch/SourceChannelImpl.java:132]: Missing import for `AIX`.

**Comment by keithc-ca** [comment]
I aborted the build; there are many errors like this:
```
[2024-09-26T15:14:13.110Z] /home/jenkins/workspace/Build_JDK17_ppc64_aix_Personal/src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java:1023: error: cannot find symbol
[2024-09-26T15:14:13.110Z]                     if (!AIX.isAIX)
[2024-09-26T15:14:13.110Z]                          ^
[2024-09-26T15:14:13.110Z]   symbol:   variable AIX
[2024-09-26T15:14:13.110Z]   location: class SocketChannelImpl
```

File: src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java
Comment (shruacha1234) [reply src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java]: @pshipton Updated. Can you please take a look

**Review by keithc-ca — CHANGES_REQUESTED** [review]
Please fix the commit message. The method name is `preClose()` (with a lowercase `p`). Lines in the body of that message should be no more than 72 characters long.

Should this not also do the same for uses of `SctpNet.preClose()`?

**Comment by pshipton** [comment]
jenkins compile aix jdk17

**Comment by pshipton** [comment]
100 iterations of java/net/Socket/asyncClose/Race.java passed https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3843/
Without this fix, this test seems to be hanging https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3844

Also running jdk_net and jdk_nio testing. We're not setup to run jdk_net testing and there are a number of failures. We'll have to determine if any are caused by this change.
https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3842

**Comment by shruacha1234** [comment]
> Please fix the commit message. The method name is `preClose()` (with a lowercase `p`). Lines in the body of that message should be no more than 72 characters long.
> 
> Should this not also do the same for uses of `SctpNet.preClose()`?

I will check on this and get back

**Comment by shruacha1234** [comment]
> Please fix the commit message. The method name is `preClose()` (with a lowercase `p`). Lines in the body of that message should be no more than 72 characters long.
> Should this not also do the same for uses of `SctpNet.preClose()`?

`SctpNet.preClose()` is used in below files
```
src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java
src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpServerChannelImpl.java
```

But we have corresponding AIX specific files where `preClose()` method is not referred there.
```
jdk.sctp/aix/classes/sun/nio/ch/sctp/SctpChannelImpl.java
jdk.sctp/aix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
jdk.sctp/aix/classes/sun/nio/ch/sctp/SctpServerChannelImpl.java
```

**Comment by keithc-ca** [comment]
> NioSocketImpl.close() will call preClose(fd) even if reader==0 && writer==0

No, `tryClose()` returns `true` if both `readerThread` and `writerThread` are zero.

**Comment by keithc-ca** [comment]
Normally, we would require changes for newer versions before merging this, but the problems this aims to fix haven't been seen in newer versions (for reasons that are not clear to me).

**Review by keithc-ca — APPROVED** [review]

**Review by pshipton — APPROVED** [review]

**Comment by mstoodle** [comment]
> > NioSocketImpl.close() will call preClose(fd) even if reader==0 && writer==0

> No, tryClose() returns true if both readerThread and writerThread are zero.

Yep. Annoying how inconsistently the logic is implemented :( . Oh well. Thanks for following up on this point.

**Comment by shruacha1234** [comment]
@pshipton
Can we please run some stress tests on this change

**Comment by pshipton** [comment]
The project isn't setup to have this sort of testing for this type of change. I ran the testing I could find before this was merged, with no concerns. It will continue to go through the nightly and weekly project testing, as well as the 0.48 release testing.

**Comment by pshipton** [comment]
@llxia is there any additional testing besides openjdk jdk_net, jdk_nio tests and the regular (unexcluded) automated tests?

**Comment by llxia** [comment]
The `jdk_nio` suite includes tests for `jdk/nio`, `java/nio`, and `sun/nio`. I couldn't think of any additional tests beyond the standard ones. As mentioned, the code will undergo our nightly, weekly, and release testing.

**Comment by shruacha1234** [comment]
@pshipton 
Could you please let me know when you expect the testing to be completed and we will get the results?

**Comment by mstoodle** [comment]
Would it make any sense to grind on some of those tests we know (hope?) exercise these code paths? I know it's not as useful as, say, those 3 hour stress tests we (used to?) run, but maybe the best we can do and maybe it's better than the number of runs in the regular testing? (lot of maybe's there, so I'll defer to Peter's, Keith's, and Lan's feelings on whether it would be worthwhile).

**Comment by mstoodle** [comment]
Ah, I hadn't caught that Peter already did a 100-run grinder here https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/386#issuecomment-2377792437 . So maybe there isn't anything more that's warranted...

**Comment by pshipton** [comment]
I also did some internal testing, and it's been going through nightly builds without issue. I'll say again though that we don't test this type of stuff well, some of it is excluded in the automated testing.

**Comment by pshipton** [comment]
We still run a 3hour test, it runs nightly. Not sure how well it tests this type of change though.

**Comment by pshipton** [comment]
For that it's worth, the 3h test is called MiniMix_3h.
https://trssrtp1.fyre.ibm.com/output/test?id=66fe6cdeb7c1433ca594d3e6

**Comment by shruacha1234** [comment]
@pshipton Can you please let us know the result of automated tests ran on weekend.

**Comment by pshipton** [comment]
jenkins restarted and the build didn't complete.

**Comment by shruacha1234** [comment]
@pshipton 
Can you please let me know on the status of the test. Are we seeing any progress in the build?

**Comment by pshipton** [comment]
There are no concerns found in the automated testing. Although we missed the last weekend build, the change is in the 0.48 release which has gone through M1 testing.

**Comment by pshipton** [comment]
BTW I expect to have the M1 build published soon, if somebody wants to pick it up and give it a try.

**Comment by pshipton** [comment]
The 0.48 M1 builds are published.
https://github.com/ibmruntimes/semeru17-binaries/releases/tag/jdk-17.0.13%2B10_openj9-0.48.0-m1

**Comment by shruacha1234** [comment]
@mstoodle 
Can we share the M1 build with the customer for validation?

**Comment by mstoodle** [comment]
> @mstoodle
> Can we share the M1 build with the customer for validation?

@shruacha1234 Sorry, I have been on vacation. I don't see why not, we produce th binaries specifically to get early feedback from our users.

**Pull request merged**
